src/psyclone/tests/psyir/backend/fortran_test.py

```diff
@@ -616,49 +590,132 @@ def test_fw_codeblock():
         "  a=1\n"
+        "  print *,\"I am a code block\"\n"
+        "  print *,\"with more than one line\"\n"
         "end subroutine tmp\n"
         "end module test")
     schedule = create_schedule(code)
 
-    code1 = (
-        "print *, 'I am a code block'\n"
-        "print *, 'with more than one line'\n")
-    _ = ParserFactory().create(std="f2003")
-    reader = get_reader(code1)
-    statements = Fortran2003.Execution_Part(reader)
-    code_block = CodeBlock([statements], parent=schedule)
-    schedule.addchild(code_block)
+    # Check a code block exists in the schedule
+    assert schedule.walk(CodeBlock)
 
     # Generate Fortran from the PSyIR schedule
-    fvisitor = FortranWriter()
-    result = fvisitor(schedule)
+    result = fort_writer(schedule)
 
     assert (
         "    a=1\n"
-        "PRINT *, 'I am a code block'\n"
-        "    PRINT *, 'with more than one line'\n" in result)
+        "    PRINT *, \"I am a code block\"\n"
+        "    PRINT *, \"with more than one line\"\n" in result)
 
 
-def test_fw_loop():
-    '''Check the FortranWriter class loop method
-    correctly prints out the Fortran representation.
+@pytest.mark.xfail(reason="issue #388. Code blocks add space and newline.")
+def test_fw_codeblock_2(fort_writer):
+    '''Check the FortranWriter class array method correctly prints out the
+    Fortran representation when there is a code block that is part of
+    a line (not a whole line). In this case the ":" in the array
+    access is a code block.
 
     '''
     # Generate fparser2 parse tree from Fortran code.
     code = (
         "module test\n"
         "contains\n"
-        "subroutine tmp()\n"
-        "  integer :: i, sum\n"
-        "  sum = 0\n"
-        "  do i = 1, 20, 2\n"
-        "    sum = sum + i\n"
-        "  end do\n"
+        "subroutine tmp(a,n)\n"
+        "  integer, intent(in) :: n\n"
+        "  real, intent(out) :: a(n,n,n)\n"
+        "    a(2,n,:) = 0.0\n"
         "end subroutine tmp\n"
         "end module test")
     schedule = create_schedule(code)
 
+    # Check a code block exists in the schedule
+    assert schedule.walk(CodeBlock)
+
     # Generate Fortran from the PSyIR schedule
-    fvisitor = FortranWriter()
-    result = fvisitor(schedule)
-    assert "do i = 1, 20, 2\n" in result
+    result = fort_writer(schedule)
+    assert "a(2,n,:)=0.0" in result
+
+
+def get_nemo_schedule(parser, code):
+    '''Utility function that returns the first schedule for a code with
+    the nemo api.
+    :param parser: the parser class.
+    :type parser: :py:class:`fparser.two.Fortran2003.Program`
+    :param str code: the code as a string.
+    :returns: the first schedule in the supplied code.
+    :rtype: :py:class:`psyclone.nemo.NemoInvokeSchedule`
+    '''
+    from psyclone.psyGen import PSyFactory
+    reader = FortranStringReader(code)
+    prog = parser(reader)
+    psy = PSyFactory(api="nemo").create(prog)
+    return psy.invokes.invoke_list[0].schedule
```

Comment:
That seems pretty useful - should this perhaps be moved into tests/psyclone_test_utils.py?

Reply:
I can't get this to work. If you move the function to tests/psyclone_test_utils.py then there is no way to get access to it as the fortran_test.py file is in a subdirectory and neither relative nor absolute imports work.

This highlights the bigger problem that tests/psyclone_test_utils functions are not available to tests in subdirectories. The only way I can see that would work without code replication would be to make psyclone_test_utils part of the PSyclone package. I think this could be another Weds discussion topic.

Reply:
This is covered in #508.